File: app/models/esi/download_my_orders.rb
require_relative 'download'

class Esi::DownloadMyOrders < Esi::Download

  def initialize( debug_request: false )
    super( nil, {}, debug_request: debug_request )
  end

  def update( user )

    if user.locked
      puts "#{user.name} is locked. Skipping ..."
      return
    end

    user_id = user.uid
    @rest_url = "characters/#{user_id}/orders/"

    return unless set_auth_token( user )

    pages = get_all_pages

    unless pages
      user.update( locked: true )
      return
    end

    current_trade_orders_id = UserSaleOrder.pluck(:id )

    ActiveRecord::Base.transaction do

      pages.each do |page|
        eve_item_id = EveItem.to_eve_item_id(page['type_id'])
        trade_hub_id =  Station.to_trade_hub_id(page['location_id'])

        # We can set orders in other hubs than those we have in the database.
        next unless trade_hub_id

        to = UserSaleOrder.where(user: user, eve_item_id: eve_item_id, trade_hub_id: trade_hub_id ).first_or_initialize
        to.price = page['price']
        to.save!

        current_trade_orders_id.delete( to.id )

        if user.remove_occuped_places
          ProductionList.where(user_id: user.id, eve_item_id: eve_item_id, trade_hub_id: trade_hub_id ).delete_all
        end
      end

      user.user_sale_orders.where( id: current_trade_orders_id ).delete_all

      user.update( download_orders_running: false, last_orders_download: Time.now )

    end
  end

end

File: esi/download_my_orders.py
import logging
from datetime import datetime

from esi.client import EsiClient
from evebs.extensions import db
from evebs.models import EveItem, UniverseStation, UniverseSystem, UserSaleOrder, ProductionList

logger = logging.getLogger(__name__)


class DownloadMyOrders:
    def update(self, user):
        if user.locked:
            logger.debug('%s is locked. Skipping.', user.name)
            return

        client = EsiClient(f'characters/{user.uid}/orders/')
        if not client.set_auth_token(user):
            return

        pages = client.get_all_pages()
        if not pages:
            user.locked = True
            db.session.commit()
            return

        print(pages)

        current_order_ids = [o.id for o in UserSaleOrder.query.filter_by(user_id=user.id).all()]

        for page in pages:
            eve_item_id = EveItem.to_eve_item_id(page['type_id'])
            station = db.session.get(UniverseStation, page['location_id'])
            if station:
                system = (UniverseSystem.query
                          .filter_by(id=station.universe_system_id)
                          .first())
                universe_system_id = system.id if system else None
            else:
                universe_system_id = None

            if not universe_system_id:
                continue

            order = UserSaleOrder.query.filter_by(
                user_id=user.id, eve_item_id=eve_item_id, universe_system_id=universe_system_id
            ).first()
            if not order:
                order = UserSaleOrder(user_id=user.id, eve_item_id=eve_item_id,
                                      universe_system_id=universe_system_id, price=page['price'])
                db.session.add(order)
                db.session.flush()
            else:
                order.price = page['price']
                if order.id in current_order_ids:
                    current_order_ids.remove(order.id)

            if user.remove_occuped_places:
                ProductionList.query.filter_by(
                    user_id=user.id, eve_item_id=eve_item_id, universe_system_id=universe_system_id
                ).delete()

        UserSaleOrder.query.filter(UserSaleOrder.id.in_(current_order_ids)).delete(
            synchronize_session=False
        )
        user.download_orders_running = False
        user.last_orders_download = datetime.utcnow()
        db.session.commit()
